File: scripts/hessenUpperSecondaryPaths.ts
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const ROOT_DIR = path.resolve(__dirname, '..')
const TOOLING_REGISTRY_PATH_CANDIDATES = [
  path.join(ROOT_DIR, 'curricula/DE/Gymnasium/input/DE-HE/retained-asset-registry.json'),
  path.join(ROOT_DIR, 'curricula/DE/Gymnasium/input/HE/retained-asset-registry.json'),
]
const LEGACY_UPPER_SECONDARY_PATH_PREFIX = 'curricula/DE/Gymnasium/input/DE-HE/'
const UPPER_SECONDARY_PATH_PREFIX = 'curricula/DE/Gymnasium/input/HE/'

function normalizeToolingPath(filePath: string): string {
  return filePath.replaceAll(
    LEGACY_UPPER_SECONDARY_PATH_PREFIX,
    UPPER_SECONDARY_PATH_PREFIX,
  )
}

function resolveToolingRegistryPath(): string {
  const pathCandidate = TOOLING_REGISTRY_PATH_CANDIDATES.find((candidate) =>
    fs.existsSync(candidate),
  )
  if (!pathCandidate) {
    throw new Error(
      `Missing Hessen upper-secondary tooling registry: ${TOOLING_REGISTRY_PATH_CANDIDATES.join(', ')}`,
    )
  }
  return pathCandidate
}

type NormalizedToolingRegistry = Omit<
  ToolingRegistry,
  'abiArchivePath' | 'mappingArchivePath' | 'sourceLandscapeRegistryPath'
> & {
  abiArchivePath: string
  mappingArchivePath: string
  sourceLandscapeRegistryPath: string
}

type ToolingSubjectEntry = {
  abiDirectory: string
  landscapeId: string
  mappingFile: string
}

type ToolingRegistry = {
  abiArchivePath: string
  mappingArchivePath: string
  sourceLandscapeRegistryPath: string
  subjects: Record<string, ToolingSubjectEntry>
}

type SourceRegistryEntry = {
  archiveSourcePath?: string
  landscapeId: string
  sourcePath: string
}

function loadJson<T>(filePath: string): T {
  return JSON.parse(fs.readFileSync(filePath, 'utf-8')) as T
}

const rawToolingRegistry = loadJson<ToolingRegistry>(resolveToolingRegistryPath())
const toolingRegistry = {
  ...rawToolingRegistry,
  abiArchivePath: normalizeToolingPath(rawToolingRegistry.abiArchivePath),
  mappingArchivePath: normalizeToolingPath(
    rawToolingRegistry.mappingArchivePath,
  ),
  sourceLandscapeRegistryPath: normalizeToolingPath(
    rawToolingRegistry.sourceLandscapeRegistryPath,
  ),
} as NormalizedToolingRegistry
const sourceRegistryPath = path.join(
  ROOT_DIR,
  toolingRegistry.sourceLandscapeRegistryPath,
)
const sourceRegistryEntries = new Map(
  loadJson<{ entries: SourceRegistryEntry[] }>(sourceRegistryPath).entries.map((entry) => [
    entry.landscapeId,
    {
      ...entry,
      archiveSourcePath: entry.archiveSourcePath
        ? normalizeToolingPath(entry.archiveSourcePath)
        : undefined,
      sourcePath: normalizeToolingPath(entry.sourcePath),
    },
  ]),
)

function getSubjectEntry(subjectKey: string): ToolingSubjectEntry {
  const subjectEntry = toolingRegistry.subjects[subjectKey]
  if (!subjectEntry) {
    throw new Error(`Unknown Hessen upper-secondary subject key: ${subjectKey}`)
  }
  return subjectEntry
}

export function resolveHessenUpperSecondaryAbiDirectory(
  subjectKey: string,
): string {
  return path.join(
    ROOT_DIR,
    toolingRegistry.abiArchivePath,
    getSubjectEntry(subjectKey).abiDirectory,
  )
}

export function resolveHessenUpperSecondaryLandscapePath(
  subjectKey: string,
): string {
  const subjectEntry = getSubjectEntry(subjectKey)
  const sourceEntry = sourceRegistryEntries.get(subjectEntry.landscapeId)
  if (!sourceEntry) {
    throw new Error(
      `Missing source-landscape registry entry for Hessen upper-secondary landscapeId ${subjectEntry.landscapeId}`,
    )
  }
  return path.join(
    ROOT_DIR,
    sourceEntry.archiveSourcePath || sourceEntry.sourcePath,
  )
}

export function resolveHessenUpperSecondaryMappingPath(
  subjectKey: string,
): string {
  return path.join(
    ROOT_DIR,
    toolingRegistry.mappingArchivePath,
    getSubjectEntry(subjectKey).mappingFile,
  )
}

File: scripts/hessen_upper_secondary_paths.py
from __future__ import annotations

import json
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path


ROOT = Path(__file__).resolve().parent.parent
TOOLING_REGISTRY_PATH_CANDIDATES = (
    ROOT / "curricula/DE/Gymnasium/input/DE-HE/retained-asset-registry.json",
    ROOT / "curricula/DE/Gymnasium/input/HE/retained-asset-registry.json",
)
LEGACY_UPPER_SECONDARY_PATH_PREFIX = "curricula/DE/Gymnasium/input/DE-HE/"
UPPER_SECONDARY_PATH_PREFIX = "curricula/DE/Gymnasium/input/HE/"


@dataclass(frozen=True)
class HessenUpperSecondaryExamPaths:
    source_landscape_id: str
    source_landscape_registry_path: Path
    landscape_path: Path
    slot_matrix_path: Path
    coverage_path: Path
    task_bank_path: Path


def _load_json(path: Path) -> dict:
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def _normalize_tooling_path(path_value: str) -> str:
    return path_value.replace(
        LEGACY_UPPER_SECONDARY_PATH_PREFIX,
        UPPER_SECONDARY_PATH_PREFIX,
    )


def _tooling_registry_path() -> Path:
    for candidate in TOOLING_REGISTRY_PATH_CANDIDATES:
        if candidate.exists():
            return candidate
    raise FileNotFoundError(
        "Missing Hessen upper-secondary tooling registry: "
        + ", ".join(str(path) for path in TOOLING_REGISTRY_PATH_CANDIDATES)
    )


@lru_cache(maxsize=1)
def _tooling_registry() -> dict:
    return _load_json(_tooling_registry_path())


@lru_cache(maxsize=1)
def _source_registry_entries() -> dict[str, dict]:
    source_registry_path = ROOT / _normalize_tooling_path(
        _tooling_registry()["sourceLandscapeRegistryPath"]
    )
    source_registry = _load_json(source_registry_path)
    return {
        entry["landscapeId"]: {
            **entry,
            "archiveSourcePath": (
                _normalize_tooling_path(entry["archiveSourcePath"])
                if entry.get("archiveSourcePath")
                else None
            ),
            "sourcePath": _normalize_tooling_path(entry["sourcePath"]),
        }
        for entry in source_registry["entries"]
    }


def _subject_registry(subject_key: str) -> dict:
    subject = _tooling_registry()["subjects"].get(subject_key)
    if subject is None:
        raise KeyError(f"Unknown Hessen upper-secondary subject key: {subject_key}")
    return subject


def resolve_hessen_upper_secondary_landscape_path(subject_key: str) -> Path:
    subject = _subject_registry(subject_key)
    landscape_entry = _source_registry_entries().get(subject["landscapeId"])
    if landscape_entry is None:
        raise KeyError(
            f"Missing source-landscape registry entry for Hessen upper-secondary landscapeId "
            f"{subject['landscapeId']}"
        )
    archive_source_path = landscape_entry.get("archiveSourcePath")
    if archive_source_path:
        return ROOT / _normalize_tooling_path(archive_source_path)
    return ROOT / _normalize_tooling_path(landscape_entry["sourcePath"])


def resolve_hessen_upper_secondary_landscape_directory(subject_key: str) -> Path:
    return resolve_hessen_upper_secondary_landscape_path(subject_key).parent


def resolve_hessen_upper_secondary_abi_directory(subject_key: str) -> Path:
    subject = _subject_registry(subject_key)
    return ROOT / _tooling_registry()["abiArchivePath"] / subject["abiDirectory"]


def resolve_hessen_upper_secondary_mapping_path(subject_key: str) -> Path:
    subject = _subject_registry(subject_key)
    mapping_file = subject.get("mappingFile")
    if not mapping_file:
        raise KeyError(f"Missing Hessen upper-secondary mapping file for subject key: {subject_key}")
    return ROOT / _normalize_tooling_path(_tooling_registry()["mappingArchivePath"]) / mapping_file


def resolve_hessen_upper_secondary_exam_paths(subject_key: str) -> HessenUpperSecondaryExamPaths:
    abi_directory = resolve_hessen_upper_secondary_abi_directory(subject_key)
    subject = _subject_registry(subject_key)
    return HessenUpperSecondaryExamPaths(
        source_landscape_id=subject["landscapeId"],
        source_landscape_registry_path=ROOT
        / _normalize_tooling_path(_tooling_registry()["sourceLandscapeRegistryPath"]),
        landscape_path=resolve_hessen_upper_secondary_landscape_path(subject_key),
        slot_matrix_path=abi_directory / "slot_matrix.json",
        coverage_path=abi_directory / "coverage_requirements.json",
        task_bank_path=abi_directory / "task_bank.json",
    )
